Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp

```diff
@@ -323,55 +323,133 @@ MLNodeLaplacian::interpolation (int amrlev, int fmglev, MultiFab& fine, const Mu
         else if (sigma[0] == nullptr)
         {
-            AMREX_HOST_DEVICE_PARALLEL_FOR_3D(bx, i, j, k,
+            experimental::ParallelFor(fine, ng,[=] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept
             {
-                mlndlap_interpadd_c(i,j,k,ffab,cfab,mfab);
+                mlndlap_interpadd_c(box_no, i, j, k, fine_ma, crse_ma, msk_ma);
             });
         }
         else if (m_use_harmonic_average && fmglev > 0)
         {
-            AMREX_D_TERM(Array4<Real const> const& sxfab = sigma[0]->const_array(mfi);,
-                         Array4<Real const> const& syfab = sigma[1]->const_array(mfi);,
-                         Array4<Real const> const& szfab = sigma[2]->const_array(mfi););
-            AMREX_HOST_DEVICE_PARALLEL_FOR_3D(bx, i, j, k,
+            AMREX_D_TERM(MultiArray4<Real const> const& sx_ma = sigma[0]->const_arrays();,
+                         MultiArray4<Real const> const& sy_ma = sigma[1]->const_arrays();,
+                         MultiArray4<Real const> const& sz_ma = sigma[2]->const_arrays(););
+            experimental::ParallelFor(fine, ng,[=] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept
+            {
+                mlndlap_interpadd_ha(box_no, i, j, k, fine_ma, crse_ma, AMREX_D_DECL(sx_ma,sy_ma,sz_ma), msk_ma);
+            });
+        }
+        else if (regular_coarsening)
+        {
+            auto sig_ma = sigma[0]->const_arrays();
+            experimental::ParallelFor(fine, ng,[=] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept
             {
-                mlndlap_interpadd_ha(i,j,k,ffab,cfab,AMREX_D_DECL(sxfab,syfab,szfab),mfab);
+                mlndlap_interpadd_aa(box_no, i, j, k, fine_ma, crse_ma, sig_ma, msk_ma);
             });
         }
         else
         {
-            Array4<Real const> const& sfab = sigma[0]->const_array(mfi);
-            if (regular_coarsening)
+            auto sig_ma = sigma[0]->const_arrays();
+            experimental::ParallelFor(fine, ng,[=] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept
             {
-                AMREX_HOST_DEVICE_PARALLEL_FOR_3D(bx, i, j, k,
+                mlndlap_semi_interpadd_aa(box_no, i, j, k, fine_ma, crse_ma, sig_ma, msk_ma, idir);
+            });
+        }
+        Gpu::synchronize();
+    } else {
+        if (m_coarsening_strategy == CoarseningStrategy::RAP)
+        {
+            auto sten_ma = stencil->const_arrays();
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+            for (MFIter mfi(fine, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+            {
+                int box_no = mfi.index();
+                Box const& amrex_i_box = mfi.tilebox();
+                LoopConcurrentOnCpu(amrex_i_box,[=] (int i, int j, int k) noexcept
                 {
-                    mlndlap_interpadd_aa(i,j,k,ffab,cfab,sfab,mfab);
+                    mlndlap_interpadd_rap(box_no, i, j, k, fine_ma, crse_ma, sten_ma, msk_ma);
                 });
```

Comment:
I suggest we use the traditional approach for cpu code.  There is also a bug here. `mfi.index()` returns the global box index, but we need to use the local box index to access the `Array4` inside `MultiArray4`. It would be simpler to just use the traditional approach.

Reply:
When I use mfi.LocalTileIndex(), the example I'm running segfaults.

Reply:
Tile index is different.  I believe `LocalIndex` is the one that could be used here.  It can be confusing.  That's why I think the traditional `MultiFab[MFIter]` approach is clearer when we are not using the MF ParallelFor.

Reply:
LocalIndex works. Thanks